quantum/process_keycode/process_joystick.c

```diff
@@ -0,0 +1,168 @@
+#include "joystick.h"
+#include "process_joystick.h"
+
+#include "analog.h"
+
+#include <string.h>
+#include <math.h>
+
+bool process_joystick_buttons(uint16_t keycode, keyrecord_t *record);
+
+bool process_joystick(uint16_t keycode, keyrecord_t *record) {
+    if (process_joystick_buttons(keycode, record) && (joystick_status.status & JS_UPDATED) > 0) {
+        send_joystick_packet(&joystick_status);
+        joystick_status.status &= ~JS_UPDATED;
+    }
+
+    return true;
+}
+
+__attribute__((weak))
+void joystick_task(void) {
+    if (process_joystick_analogread() && (joystick_status.status & JS_UPDATED)) {
+        send_joystick_packet(&joystick_status);
+        joystick_status.status &= ~JS_UPDATED;
+    }
+}
+
+bool process_joystick_buttons(uint16_t keycode, keyrecord_t *record) {
+    if (keycode < JS_BUTTON0 || keycode > JS_BUTTON_MAX) {
+        return true;
+    } else {
+        if (record->event.pressed) {
+            joystick_status.buttons[(keycode - JS_BUTTON0) / 8] |= 1 << (keycode % 8);
+        } else {
+            joystick_status.buttons[(keycode - JS_BUTTON0) / 8] &= ~(1 << (keycode % 8));
+        }
+
+        joystick_status.status |= JS_UPDATED;
+    }
+
+    return true;
+}
+
+uint16_t savePinState(pin_t pin) {
+#ifdef __AVR__
+    uint8_t pinNumber = pin & 0xF;
+    return ((PORTx_ADDRESS(pin) >> pinNumber) & 0x1) << 1 | ((DDRx_ADDRESS(pin) >> pinNumber) & 0x1);
+#elif defined(PROTOCOL_CHIBIOS)
+    /*
+    The pin configuration is backed up in the following format : 
+ bit  15    9  8   7   6  5  4   3     2    1 0
+      |unused|ODR|IDR|PUPDR|OSPEEDR|OTYPER|MODER|
+    */
+    return  (( PAL_PORT(pin)->MODER   >> (2*PAL_PAD(pin))) & 0x3)
+          | (((PAL_PORT(pin)->OTYPER  >> (1*PAL_PAD(pin))) & 0x1) << 2) 
+          | (((PAL_PORT(pin)->OSPEEDR >> (2*PAL_PAD(pin))) & 0x3) << 3)
+          | (((PAL_PORT(pin)->PUPDR   >> (2*PAL_PAD(pin))) & 0x3) << 5)
+          | (((PAL_PORT(pin)->IDR     >> (1*PAL_PAD(pin))) & 0x1) << 7)
+          | (((PAL_PORT(pin)->ODR     >> (1*PAL_PAD(pin))) & 0x1) << 8);
```

Comment:
This code is problematic — it is not just ChibiOS-specific, but works only with some particular STM32 chips (e.g., it does not compile for `handwired/onekey/bluepill:joystick`, which has STM32F103 with a different GPIO hardware block). As a minimum, it should have some hardware-specific `#ifdef`s, so that unsupported chips could still support a virtual joystick.